### ISSUE TITLE
New fragmentation run

### ISSUE THREAD

**Issue by Waztom** [m2ms/fragalysis-frontend#1486]
Matteo is providing a new set of molecules that need to be added to the graph database. The expectation is that there will be about 200M molecules.
To process this we need to re-instate the fragmentation machinery and as the database will now not be co-located with the compute cluster we need top make minor changes to the process to copy files between clusters.

We aim to process this data as a new dataset (e.g. starting with an empty database) and then use the combine play to combine it with the old data. Hence the process will look like this:
1. standardize the new molecules (cluster)
2. load standardized molecules into the database (database)
3. extract out the molecules to be fragmented (database)
4. fragment (cluster)
5. load fragmented data into database (database)
6. extract out molecules needing generation of additional info (e.g. InChi) (database)
7. generate additional info (cluster)
8. load additional info into database (database)
9. generate nodes and edges csv files (databse)
10. combine with existing data (cluster)

We plan to run a small test dataset through this process to make sure that it's running properly.

**Comment by mwinokan** [comment]
Chris Reynolds is able to assist after shutdown (after week 1 of Sept).

@Waztom asks if we can move this job to the DLS cluster (not STFC). Either way @tdudgeon says we need significant resources, which are available at DLS (150TB of GPFS).

@tdudgeon assumes that functionality the DLS cluster will be the same as STFC as they both use SLURM. @Waztom says we can leverage @ConorFWild's experience rather than relying on Chris.

Object storage for IRIS/STFC has not been investigated by @Waztom or @mwinokan.

The CPU requirement is around 2000 cores. @mwinokan has briefly checked and there are around 56 nodes with 64 cores and 500GB of RAM idle, both `gpfs03` and `gpfs04` filesystems are mounted.

The PostGres server will need to be available as well.

**Comment by alanbchristie** [comment]
A simple diagram illustrating non-cluster elements (like DB and NFS server) along with the "expected" shared filesystem: -



![Image](https://github.com/user-attachments/assets/82df63e7-1f90-4038-8e78-ca14b19130ba)

**Comment by mwinokan** [comment]
@ConorFWild says that obtaining 2000 concurrently available cores will need SC to increase the job limits. Graham is the SC contact for this

**Comment by alanbchristie** [comment]
SC Request ID: **SCHD-5779**

**Comment by alanbchristie** [comment]
The SC Request has been shutdown and closed as **"Won't Do"**.

I have forwarded the email but clearly they is no desire to support execution outside of Iris. I will step aside on this topic as there is nothing more I can do.

**Comment by mwinokan** [comment]
@alanbchristie has been testing with a smaller (1M) dataset and optimising parameters. We are ready to fragment a bigger dataset

**Comment by tdudgeon** [comment]
Here are some example molecules that are currently excluded from fragmentation (heavy atom count > 36 or num frags > 12)

![Image](https://github.com/user-attachments/assets/ce73cbcf-2a05-4739-805a-2197296305fa)

The good news is that file is already being created, the bad news is that the IDs are not being written, so we need to do a small tweak to allow it to be used as an input.

**Comment by phraenquex** [comment]
@tdudgeon - great, we definitely don't need that shit for our scaffolds.  

(@mwinokan @Waztom do say if you disagree.)

**Comment by matteoferla** [comment]
Sorry all, I should have nudged my query email ages back to get the details straight on the onset.
@tdudgeon, What does `num frags > 12` mean? Are those BRICS decompositions?

Re the screenshot. There was a penalty for excessive methylene groups, but not for methyl groups.
Triazoles and tetrazoles are _uncommon_ in Enamine, but triazoles (Huisgen cycloaddition both 1,3 = Cu and 1,5 = Ru) were enriched by the synthon enrichment used [custom SMARTS patterns](https://github.com/matteoferla/Poised-Fragment-library-design/blob/main/library_subsetting/library_subsetting_v3/data/decompositions.yaml) for common late stage functionalisation reaction products.
Re morpholino groups, hetero-alicyclics such as were marginally favoured by a filtered aimed at limiting polyphenyl chains.
Majorly, the molecules are rich in HBond donors and have high TPSA. Whereas these are denigrated as bad for permeability, the compounds that bind frequently in XChem have high TPSA and number of enthalpic interactions, whereas entrophic / hydrophobic / greasy interactions underperform both in the screen and as followups reflecting the difference between what is best for the crystallographic screen (=well placed compounds) vs. what is best for cell based assays. Consequently, the strongly tailed lower quartile of the greasy Enamine compounds was excluded, but not an upper quantile of high TSPA, which are already filtered by Enamine.

**Comment by tdudgeon** [comment]
`num frags` is the number of fragments that would be generated by the fragment network

**Comment by phraenquex** [comment]
Thanks @matteoferla - the relevant question at this point, is whether they are likely to make "good" scaffolds, i.e. very productive input for Syndirella.  

And to summarise what we arrived at yesterday:  if the answer is only "_maybe_" as well as "_but they'll take awfully long to fragment and slow down Max getting a functional Syndirnet_", then they get thrown onto a "don't do now" pile, and we'll deal with them later, e.g. as a slow-running low-priority fragmentation queue.

**Comment by phraenquex** [comment]
Transferring an email thread for the record - especially as it slightly documents (a bit) the algorithm.

@matteoferla Sept 14: 

> I filtered down the 7B molecule catalogues in three steps, followed by a final sort.
>`6B ==first and second pass==> 220M ==third pass==> 20M ==sorting==> ordered 20M.`
> The number of compounds that can be fit in the network I am told is variable in the 10M-20M range, hence the ordered part.
>
> (The order comes from a sum of weighted Winsorised Z-scores, accounting for pharmacophoric trio distance uniqueness, presence of reaction product moieties, presence/absence of certain moieties (e.g. excessive number of benzenes and alkanes) and regular filters, all with a bonus for those that were a partial superstructure of a XChem library compound, in accordance to Frank’s request)
>
> I sent you a link to a MS OneDrive folder with the ordered 20M compounds. This was not my ideal choice, but GitHub was being problematic —hence my delay.
>
> Warren tells me you might have access to the NFS already, in which case
>
> The data is in the ‘shared XChem drive’, namely the NTF folder mounted from 192.168.212.119:/mnt/xchem-fragalysis-2  as /opt/xchem-fragalysis-2 in some locations. Let’s say this is $DATA In the folder `$DATA/mferla/library_making/selected_final` contains ordered smiles as mentioned.
>
> In the folders in $DATA/mferla/library_making/third_pass are the conformers. These are clustered by the sorting metric in Z1 (top), Z1-08 (high), Z05-08 (mid) and Z0-05 (low). Within these folder they are not sorted and are split into the files were they came from. Namely, the original Enamine files were split into chunks (eg. Enamine_REAL_HAC_29_38_1.3B_Part_2_chunk66.sdf.bz2 is the 66th 200M chunk of REAL HAC 29-38).
>
> $DATA/mferla/library_making/second_pass is smiles and is with basic filters, without sorting for pharmacophoric and synthetic properties. It is not needed (it’s the 220M), but I for one would be curious as to what is in ‘2’ after ‘3’ gets mentioned.
>
> Unfortunately, just now I noticed that the HAC27 subset of REAL is missing. I will run it over the weekend and will be in third_pass folder,
>
> But the ordered set on OneDrive will need deleting and remaking, which is a problem as I will be travelling on Monday.

**Comment by phraenquex** [comment]
@alanbchristie 19 Sept: 

> It’s taken more time than I expected to get to initial processing of the data. One thing that does stand out is the dramatic change to the identifier style throughout the file. In the early part of the file it’s a pleasant-looking MCULE-2832428593. Deeper into the file I see things like m_274552____23520642____24972820____13454082. Is the latter identity OK for users?

**Comment by phraenquex** [comment]
@tdudgeon 24 Sept:

> I just wanted to clarify which set of molecules we should be fragmenting.
> 
> The files in /opt/xchem-fragalysis-2/mferla/library_making/selected_final seem to add up to about 20M molecules (but the file shorlist0017.1M.cxsmiles.bz2 seems to be corrupt).
> 
> We were under the impression that we were to be fragmenting about 200M molecules. Is it those in /opt/xchem-fragalysis-2/mferla/library_making/second_pass that we should be using?
>
> Also, as already discussed, there are a significant number of duplicates (same SMILES but with different IDs). Is this because you are doing a desalting operation or something similar. If so it looks like the original molecule has been lost.
> 
> Not really a problem, just need to know what's going on.

**Comment by phraenquex** [comment]
@tdudgeon 25 Sept: 

> Further to this, tube duplicates DO cause a problem.
> 
> I'm assuming it's impractical for Matteo to regenerate without duplicates, so we should generate a simple preprocessing script that removes them from the data.
> 
> Presumably it's OK to just keep the first (based on SMILES) and discard any subsequent ones?

**Comment by phraenquex** [comment]
@matteoferla 25 Sept: 

> OK to just keep the first  — When ordering from Enamine, they actually disregard the ID given and may deliver with a synonym ID, so it is unnecessary to keep the synonym codes.
> 
 > No, regeneration with removal of duplicates will not result in changes in threshholds, so would be the same as removal of duplicates from the second pass set. If needed I can make a 2.1 set without duplicates —I assume the easiest way would be bin the molecules by elemental composition and from within do a pairwise comparison.

**Comment by tdudgeon** [comment]
I'm still not sure what data we are supposed to be fragmenting.
I've looked at the data in `/opt/xchem-fragalysis-2/mferla/library_making/second_pass` and I created a simple deduplication process. There were only duplicates in one set (the HAC 25 ones), but the process was useful as it allowed to remove headers which were repeated within the individual files, and to roll up the multiple files into a more manageable set.
But on looking at these there seem to be far too many molecules. I was expecting around 200M but there are approx 1.23 billion:
```
$ wc -l dedup_Enamine_REAL_HAC_*
   141887137 dedup_Enamine_REAL_HAC_11_21.cxsmiles
   130442032 dedup_Enamine_REAL_HAC_22_23.cxsmiles
    89973092 dedup_Enamine_REAL_HAC_24.cxsmiles
   181467620 dedup_Enamine_REAL_HAC_25.cxsmiles
   125579937 dedup_Enamine_REAL_HAC_26.cxsmiles
   131331981 dedup_Enamine_REAL_HAC_27.cxsmiles
   132143780 dedup_Enamine_REAL_HAC_28.cxsmiles
   299480349 dedup_Enamine_REAL_HAC_29.cxsmiles
  1232305928 total
```

So either these are not the right files, or there are far more moleucles than expected, and too many to be handled by fragmentation.

@matteoferla can you shed some light on this?

**Comment by matteoferla** [comment]
The 220M was a projection based on file sizes. So your 1B is correct.
This means that the third step is still needed, but the top quantile to keep needs relaxing as it is too strict (as it only gave 20M). The lower quantiles were not kepts as I was hoping to hear back if conformer were preferred to SMILES —which I think they aren't. This means I need to re-run the third step.

For reference, the full algorithmic details are in the repo: https://github.com/matteoferla/Poised-Fragment-library-design

**Comment by tdudgeon** [comment]
OK, we'll work with the `selected_final` set for now.
Bear in mind that the `shorlist0017.1M.cxsmiles.bz2` seems to be corrupt.
This leads to 18,894,798 molecules to process, with 105,203 excluded as duplicates.

**Comment by mwinokan** [comment]
@alanbchristie reports that the 18M run was underway, but needed to be restarted because the time estimation was wrong and it was going to time out. It should be finished on Saturday using around 560 cores.

Alan is testing how the limit to the number of fragments per compound is making the fragmentation time more consistent. For the 1M compound dataset it was a 70% reduction.

@matteoferla has kicked off another job to prepare the dataset that will be added in incrementally once he's back online

**Comment by matteoferla** [comment]
To clarify, it's a set of jobs, but the cluster is in heavy use (by someone's jobs that will be done within 72 hours, so I will have the data updated with 4-5 days

**Comment by mwinokan** [comment]
@alanbchristie reports:

- With a fragment limit of 8, it took 63 hours to process 16.6M compounds across four steps (30hours for fragmentation, 17hours for neo4j insertion, 17hours for inchikey generation).
- Only 10 hours (subset of fragmentation) are relying on a large number of parallel cores on the cluster
- @phraenquex asks if more steps of the process can be parallelised
- @alanbchristie says that the graph database might need more RAM or cores to improve performance, and there is not much scope for more resources

**Comment by alanbchristie** [comment]
The latest experimental run exposed a 32-bit table indexing problem ... i.e. we've hit more than 2 billion records during fragmentation. We've adjusted the table definitions to use `bigint` in the fragmentation image now but this has had a knock-on effect and now exposes issues building the image (which is based on a soon-to-be-out-of-date Python ... i.e. 3.8)! The CI build fails but is fixed by moving from **3.8.6-buster** (Debian 10) to **3.8.20-bullseye** (Debian 11). I suspect buster's no longer supported?

Hopefully this will have no effect on the behaviour of the image.

Heads up for any Python 3.8 images - Python 3.8 is due to go **END OF LIFE** this month (October 2024)!

**Comment by matteoferla** [comment]
@alanbchristie - A _potential_ heads up (as I have been bitten hard by it). With Singularity (no idea with Docker), there is a troublesome issue with Ubuntu 22 or later (Debian 12 based), wherein the usual song-and-dance of  `apt-get update -y` and  `apt-get upgrade -y` fails when user-permissions are restricted resulting in a dearth of images based on Ubu22 or Ubu24. Regarding the obsolescence of 3.8, there are a few modules already that hate it: IIRC torch 2 secretly does not like it at all.

**Comment by mwinokan** [comment]
@tdudgeon and @alanbchristie have received the data and has broken it into five 50M batches. The first batch of fragmentation is currently underway (will get an idea of performance over the next few days)

**Comment by tdudgeon** [comment]
From Matteo:

> The dataset filtering is all correct on my side, but my cluster usage is now limited to 3 nodes, so my 3 day estimate was crippled yesterday —I only noticed now the cause of the delay at lunchtime.
> 
> I had corrected the mistake that meant `finalists` folder causing it to be only 20M and now I have the expected number of compounds, and I have scripts in place to sort, merge and duplicate the data for sharing, but 5% are still to be processed.
> 
> In the folder 192.168.212.119:/mnt/xchem-fragalysis-2/mferla/library_making/ filtered_additional_finalists are 100M more compounds from the top tier that have been combined, sorted and filtered for duplicates —lacking 5%. Where there no HPC queuing, I would wait to email, but I just know that things will be worse. The sorted nature this set means I will make a second folder with the missing top tier compounds from the missing sets (~5M compounds) and a third compound with all the merged and deduplicated second tier compounds —which _need_ to be sorted as I was to cut it off at 100M.
>  
> There are likely duplicates as I took some shortcuts. As mentioned, compounds with multiple Late Stage Functionalisation synthetic routes to make them will have synonyms and these are enriched on purposed in this dataset.

and:

> It took a bit longer than I had planned, but in the folder library_making/ filtered_additional_finalists are 270M compounds in sequential order in files of 1M each. The file #142 is smaller because of how I processed the data. 
> 
> There is a typo in the file names, missing T, which I did not correct as that typo was present in the files in the folder selected_final (1–16).

**Comment by tdudgeon** [comment]
I looked at the data.
Again, the last file is corrupt.
In the remaining ones I find 251,755,781 molecules with 144018 duplicates leaving a total of 251,611,763 unique molecules.

**Comment by alanbchristie** [comment]
The initial fragmentation run has failed (**No space left on device**). Unfortunately we were initially out of "quota" to create a larger volume. When we did make space available we then find that we cannot attach the volume to the NFS server ... and so are waiting for a response from STFC.

**Comment by mwinokan** [comment]
@alanbchristie says that he's still waiting on STFC to attach the newly created volume to the NFS server

The ~300TB available on IRIS nodes is not available from the worker node, so it's not compatible with our workflow (and it's not clear whether Chris wants us to use this).

**Comment by mwinokan** [comment]
@tdudgeon says this is back underway and the first 50M were fragmented in around 75h (including 18h of cluster processing)

**Comment by mwinokan** [comment]
@alanbchristie says the fragmentation run is ongoing again (after a pause due to the cluster outages)

**Comment by mwinokan** [comment]
@alanbchristie says that this has stalled due to the cluster instability

**Comment by mwinokan** [comment]
@tdudgeon reports that @alanbchristie has restarted the run

**Comment by alanbchristie** [comment]
The first 250M molecules have been fragmented and extracted to Echo/S3. Here's a summary of execution stats for the `250M-C1` run: -

1. Standardisation: **2h 15m**
2. Fragmentation: **75h 30m**
3. Inchi generation: **15h 00m**
4. Extract CSV: **78h 15m**

A total run-time of about **171h** (7 days 21 hours) - i.e. `approximately 8 days` with an extracted set of CSV files totalling `72.688 GiB`

> It is important to realise that the total run-time may be longer (or shorter) than 8 days because a) each molecule set is different, and b) a human has to start each step once the previous has completed. So if a run completes at 1am the next step may not be started until maybe 8 or 9am that day (adding 8 hours to the run). So, to be conservative, add at least 8 hours to each step, e.g. 24 hours for the above, making the total run-time a more realistic `9 days`

**Comment by mwinokan** [comment]
@tdudgeon says the first chunk of 50M (`250M-C1`) has finished in about 8 days.

@alanbchristie says there was an issue on Sunday night causing the second run to need restarting, but it is now running.

**Comment by mwinokan** [comment]
@alanbchristie suspects a problem and will investigate. Please update in this ticket if anything noteworthy comes up

**Comment by mwinokan** [comment]
@alanbchristie reports that the runs stall, a few things have been tried but Alan has not found a fix yet.

**Comment by mwinokan** [comment]
@alanbchristie reports that the fragmentation problem has been resolved and the second 50M batch is nearing completion. 

@alanbchristie will then move the fragmentation job to the new cluster to free up hardware for #1451

**Comment by alanbchristie** [comment]
Fragmentation has been successfully moved to the new (k8s 1.30) cluster, and _appears_ to be running as expected. This leaves the old (k8s 1.23) cluster running the **extract** of the `C2` run. When this finishes the old cluster will be free to remove (after backing up the moldb database and old AWX data).

`C2` **extract** is _extecped_ to complete sometime tomorrow.

The new cluster is fragmenting the 50M molecule from chunk `C3` . this has already failed 3 times after many hours of execution due to unknown cluster/networking issues.

`C3` failed at `00:33:31` on Sunday night and at exactly the same time last night (25th Nov). I have moved to a new **nextflow** which supports parameters that might help with the overnight (network) failures. A 4th execution of the `C3` **fragment** run is now underway.

**Comment by phraenquex** [comment]
Thanks @alanbchristie for the super informative breakdown.

How well could the long jobs this parallelise?  Would it shrink once you engineer use of the Diamond's (much more) infinite-ish CPUs available?

**Comment by alanbchristie** [comment]
The actual 'fragment' generation in fragmentation is executed in parallel so that really just depends on cores but a significant majority of work involves database interactions some of which are run in parallel. Others might benefit from a re-write.

There's also the 'quality' of the hardware. For example I have had to start the third (C3) fragmentation run four times since last Friday after each attempt has been affected by some sort of underlying cloud/network issue.

> With regard to the 200M I am hoping we have enough time left to run the C3, C4 and C5 "chunks". C3 is already a day into its 3 day fragment run. But, if we run continue to be disrupted by the hardware we will not have enough time. I think we're looking at 24 days still to go and It's nearly December.

**Comment by phraenquex** [comment]
Thanks.  Run as much as you can, and whatever remains just doesn't make it in, no disaster.)

(Scientifically, it's precisely equivalent to Matteo having run some filter more stringently, or the vendors having a slightly smaller catalogue.  Whatever.)

When you get to reengineering it (for future iterations):  
* how can DB interactions be scaled.  (DB on larger node?)
* more robust handling of failures by dispatcher mechanism.

**Comment by mwinokan** [comment]
@alanbchristie had to restart the run after the NFS server outage. The third batch is running again now

**Comment by mwinokan** [comment]
@tdudgeon says that the database itself is probably at the peak of it's performance, even if allocated more resources. The parallel data manipulation in postgres is the limiting factor.

Tim suggests that a refactor Kafka message queues to improve the fragmentation performance beyond estimated 2x from allocating more resources to the DB, but this is a large body of work

**Comment by mwinokan** [comment]
@alanbchristie says that we are hitting networking issues but the job is running again

**Comment by alanbchristie** [comment]
The extract of the 3rd 'chunk' of 50 million molecules (C3) is nearly complete (with just the extract play to run), here's a brief summary: -

1. Standardisation: **1h 52m**
2. Fragmentation: **61h 28m**
3. Inchi generation: **17h 05m**
4. Extract CSV: **In Progress**

**Comment by mwinokan** [comment]
@alanbchristie reports that a new failure happened over the weekend, but the playbooks have been patched to allow for faster/easier restarts

**Comment by alanbchristie** [comment]
The extract of the 4th 'chunk' of 50 million molecules (C4) is complete. Here's a brief summary: -

1. Standardisation: (about) **1h 45m**
2. Fragmentation: (about) **60h**
3. Inchi generation: **22h 28m**
4. Extract CSV: **54h 18m** (to dls-echo:/im-fragnet/extract/enamine_mferla/250M-C4)

A total run-time of about 138h (5 days) - with an extracted set of CSV files totalling 92 GiB

**Comment by alanbchristie** [comment]
The extract of the 5th 'chunk' of 50 million molecules (C5) is complete. Here's a brief summary: -

1. Standardisation: (about) 1h 45m
2. Fragmentation: (about) 80h 26m
3. Inchi generation: 24h 23m
4. Extract CSV: 58h 54m (to dls-echo:/im-fragnet/extract/enamine_mferla/250M-C5)

A total run-time of about 166h (7 days) - with an extracted set of CSV files totalling 96 GiB

**Comment by alanbchristie** [comment]
- Approximate (average) processing time (per 50 million chunk): **146h (6 days)**
- Compressed extracted output (all 5 chunks): - **407Gi**

**Comment by mwinokan** [comment]
@alanbchristie is working on a dry-run of the combination step (ETA today) to get an estimate of the resources required on IRIS

@Waztom suggests aggregating on molecular weight, but @alanbchristie says that it would need changes to the playbooks and @tdudgeon says it may duplicate the existing hashing

**Comment by mwinokan** [comment]
@tdudgeon has been exploring the molecular weight binning for de-duplication.

**Comment by mwinokan** [comment]
@tdudgeon is prototyping new combine step and will work with @alanbchristie on the implementation tomorrow

**Comment by Waztom** [comment]
@tdudgeon has found a data issue with missing Inchi data.

**Comment by alanbchristie** [comment]
For historical reference, here is a compressed set of the **stdout** for each Ansible play for each of the chunks (1-5), including some failures. The archive contains the stdout for the plays: -

- standardise
- fragment
- inchi
- extract

> We had a lot of hardware-based execution problems with some of the stdout captured here (specifically failures during c3, and c4 processing)

[ansible-stdout-c1-c5-archive.zip](https://github.com/user-attachments/files/18844963/ansible-stdout-c1-c5-archive.zip)

**Comment by Waztom** [comment]
@phraenquex says delete the datasets that have missing Inchis. Frank wonders why we need to spend time on capturing these datasets; what is the scientific need for capturing all these datasets? Asks if Neo4J can deal with duplicate nodes/deal with this - one with and the other without an Inchi?  

@tdudgeon has completed de-duplication of nodes. The edge de-duplication is pending and will cause issues when trying to connect with nodes with missing Inchis.        

@alanbchristie says there are Neo4J options to skip instances when node missing info or when dealing with duplicate nodes. @phraenquex asks why we don't just use Neo4J to deal with duplication? @alanbchristie says this is an option and will test this.

**Comment by Waztom** [comment]
@alanbchristie tested two approaches:
1. Use Neo4J de-duplication/detecting bad node. This takes everything (the multiple batches) and throws it at Neo4J to sort out. Took approx. 12 hours. 
2.Concatenate all the batches in one go and throw at Neo4J - this is the most brute force test possible. This will test how much we can throw at Neo4J to handle. 

@tdudgeon is looking at an alternative approach (will not be invoiced) to de-duplication to clean up the batches pre Neo4J treatment. This requires a correction to the scripts to catch and create missing Inchi keys.

**Comment by alanbchristie** [comment]
Importing of the combination of C1 + C2 as a test to investigate what is actually logged when the following additional options are used with `/var/lib/neo4j/bin/neo4j-admin import`: -

```
--verbose
--ignore-empty-strings
--ignore-extra-columns
--skip-duplicate-nodes
--skip-bad-relationships
```

Other interesting options (unused in this experiment) are: -

```
--max-memory (default 90%)
--cache-on-heap
--bad-tolerance (1000 bad relationships)
--skip-bad-entries-logging (Whether or not to skip logging bad entries detected during import)
```

But there's not indication that logging of duplicate nodes can be stopped.

Anyway ... with C1 + C2 we find the following information: -

---

## Summary
- Import time: **15h 40m**
- Database (/data) size: **1.8TB**
- Size of `import.report`: **24.38GB** (!)
- Lines in `import.report`: **314 Million** (314,057,734)

## Import stdout
```
Importing the contents of these files into /data/data/databases/graph.db:
Nodes:
  /data/data-loader/header-inchi-nodes.csv
  /data/data-loader/inchi-nodes.csv.gz
  /data/data-loader/header-isomol-nodes.csv
  /data/data-loader/isomol-nodes.csv.gz
  /data/data-loader/header-nodes.csv
  /data/data-loader/nodes.csv.gz
  /data/data-loader/header-suppliermol-nodes.csv
  /data/data-loader/suppliermol-nodes.csv.gz
  /data/data-loader/header-supplier-nodes.csv
  /data/data-loader/supplier-nodes.csv.gz

Relationships:
  /data/data-loader/header-edges.csv
  /data/data-loader/edges.csv.gz
  /data/data-loader/header-isomol-molecule-edges.csv
  /data/data-loader/isomol-molecule-edges.csv.gz
  /data/data-loader/header-isomol-suppliermol-edges.csv
  /data/data-loader/isomol-suppliermol-edges.csv.gz
  /data/data-loader/header-molecule-inchi-edges.csv
  /data/data-loader/molecule-inchi-edges.csv.gz
  /data/data-loader/header-molecule-suppliermol-edges.csv
  /data/data-loader/molecule-suppliermol-edges.csv.gz
  /data/data-loader/header-suppliermol-supplier-edges.csv
  /data/data-loader/suppliermol-supplier-edges.csv.gz


Available resources:
  Total machine memory: 220.0GiB
  Free machine memory: 217.7GiB
  Max heap memory : 32.00GiB
  Processors: 4
  Configured max memory: 169.2GiB
  High-IO: true

Type normalization:
  Property type of 'hac' normalized from 'int' --> 'long' in /data/data-loader/header-nodes.csv
  Property type of 'chac' normalized from 'int' --> 'long' in /data/data-loader/header-nodes.csv
  Property type of 'build_number' normalized from 'int' --> 'long' in /data/data-loader/header-supplier-nodes.csv
  Property type of 'limit' normalized from 'int' --> 'long' in /data/data-loader/header-supplier-nodes.csv
  Property type of 'min_hac' normalized from 'int' --> 'long' in /data/data-loader/header-supplier-nodes.csv
  Property type of 'max_hac' normalized from 'int' --> 'long' in /data/data-loader/header-supplier-nodes.csv
Nodes, started 2025-02-25 10:50:24.571+0000
Done in 45m 56s 197ms
Prepare node index, started 2025-02-25 11:36:20.823+0000
Done in 1h 56m 33s 195ms
```

## Head of /data/data-loader/import.report
```
Id 'S' is defined more than once in group 'F2'
Id 'F' is defined more than once in group 'F2'
Id 'O' is defined more than once in group 'F2'
Id 'N' is defined more than once in group 'F2'
Id 'C' is defined more than once in group 'F2'
Id 'I' is defined more than once in group 'F2'
Id 'Br' is defined more than once in group 'F2'
Id 'NO' is defined more than once in group 'F2'
Id 'CS' is defined more than once in group 'F2'
Id 'CF' is defined more than once in group 'F2'
```

## Tail of /data/data-loader/import.report
```
Id 'CC(C)(C)Oc1ccccc1CNC=O.Cn1ccnn1' is defined more than once in group 'F2'
Id 'CC12CC3CC(C)(C1)CC(C(=O)NS(=O)(=O)c1ccc4[nH]c(=O)[nH]c4c1)(C3)C2' is defined more than once in group 'F2'
Id 'CC(c1c[nH]nn1)N(C)C(=O)c1cc([N+](=O)[O-])c[nH]1.c1ccc2[nH]cnc2c1' is defined more than once in group 'F2'
Id 'O=C(Nc1ccc2[nH]c(=O)[nH]c2c1)c1c[nH]c(-c2cccc([N+](=O)[O-])c2)n1' is defined more than once in group 'F2'
Id 'Cc1[nH]c(CN(C)c2ccc3[nH]c(=O)[nH]c3n2)nc1-c1cccc([N+](=O)[O-])c1' is defined more than once in group 'F2'
Id 'O=C(c1c[nH]c2ccc([N+](=O)[O-])cc12)N1CCCC1.c1c[nH]nn1.c1cn[nH]c1' is defined more than once in group 'F2'
Id 'CC12CC3(C)CC(O)(C1)CC(C(=O)NS(=O)(=O)c1cc([N+](=O)[O-])cs1)(C2)C3' is defined more than once in group 'F2'
Id 'O=C(Nc1ccc2[nH]c(=O)[nH]c2c1)c1cc(-c2ccc([N+](=O)[O-])cc2)n[nH]1' is defined more than once in group 'F2'
Id 'CC(=Cc1cn[nH]c1)C(=O)NC1CC2CC1N(C(=O)c1cc([N+](=O)[O-])c[nH]1)C2' is defined more than once in group 'F2'
Id 'CC(=Cc1cn[nH]c1)C(=O)N1CC2CCC(C1)N2C(=O)c1ccc([N+](=O)[O-])[nH]1' is defined more than once in group 'F2'
```

**Comment by alanbchristie** [comment]
The description here appears untrue, with over 300 million lines in the report relating to duplicate nodes: -

```
--skip-duplicate-nodes[=<true/false>]
    Whether or not to skip importing nodes that have
    the same id/group. In the event of multiple
    nodes within the same group having the same id,
    the first encountered will be imported whereas
    consecutive such nodes will be skipped. Skipped
    nodes will be logged, containing at most number
    of entities specified by bad-tolerance, unless
    otherwise specified by skip-bad-entries-logging
    option.
    Default: false
```

For the next run it is worth experimenting with: -

- --skip-bad-entries-logging
- --cache-on-heap

**Comment by alanbchristie** [comment]
The `skip-bad-entries-logging` and `cache-on-heap` import is running in the `graph-x` Namespace now...

**Comment by mwinokan** [comment]
@tdudgeon will need to rework the hashing code (fairly simple fix) and then can start deduplicate the nodes later today. Hopefully, by end of next week the new data (from Matteo) the nodes and edges will be deduplicated. Then the same will need to be performed with the existing graph

**Comment by mwinokan** [comment]
@tdudgeon updates:

- Node deduplication is nearly complete (ETA: EOD)
- Edge deduplication is underway, hashing expected to finish today
- Other data files may need deduplicating (e.g. supplier data, inchi info)
- The old data (existing graph) will need work as well, potentially missing labels and incorrect inchi's

@tdudgeon is asking whether to reprocess the old data with the new mechanisms, or to try to sanitise the existing database

@tdudgeon says that the issues manifest in missing data, as two records are created when a molecule is a fragment of a larger one and is available to purchase and one is thrown away during deduplication. The size of the affected data is around 1% of the total database, but of the purchasable subset it may be over 50%.

**Comment by mwinokan** [comment]
@tdudgeon reports that this is progressing slowly due to the size of the data. The data files for 250M from Matteo + original xchem data will be ready by Monday, and then Neo4J will need to be initiated (circa one day)

**Comment by mwinokan** [comment]
@tdudgeon reports that two sets of data have been created:

- combined diamond libraries
- combined diamond libraries w/ matteo's new compounds

The smaller set loaded into neo4j was fine. Pending attempt with the bigger set. @tdudgeon warns that this is a manual process so some human errors could creep in.

@phraenquex stresses that the automation of the fragmentation process is out of scope for now

Documentation ticket; https://github.com/m2ms/fragalysis-frontend/issues/1619

@phraenquex confirms that documentation should be made up to date with the current process, and future work

@tdudgeon asks if the chemspace / enamine (~300M) libraries in the old graph should be added into the new graph. @phraenquex says as long as the process can happen in the background and it is not a lot of human effort to monitor, then it should be combined. @phraenquex advises to start with chemspace

**Comment by mwinokan** [comment]
@tdudgeon reports that there has been an issue with the nodes in chunk 5 of the Matteo data (molecules ended up in the wrong slice). Tim has re-run the data and it is correct, recombining is in progress. @alanbchristie will check the Matteo data later today.

@tdudgeon says the full data could be ready by Monday. The chemspace library (20M) is ready as well

**Comment by mwinokan** [comment]
@tdudgeon says that we're pretty much ready, @alanbchristie is preparing the graph for the Matteo data (still loading). Once that's loaded it will be deleted and then all the data will be loaded and combined. Expected that by end of the week the XChem libraries and 250M Matteo compounds will be available.

Then can proceed with #1720

**Comment by mwinokan** [comment]
@tdudgeon says this is on track. The new new graph DB is being fired up.

@tdudgeon raises concerns that the new big dataset is much larger [432GB] than before [192GB] (more nodes and edges than expected for this quantity of molecules). 

@tdudgeon is combining the new big dataset with chemspace (complete early next week)

@Waztom asks how much more time will be spent on this topic. @tdudgeon says it depends on whether the more complex 60M molecules will need to be fragmented, and whether the old Enamine data should be included.

@phraenquex and @mwinokan stress that moving on to #1720 needs to happen to we can inform decisions on whether extra data is needed

**Comment by mwinokan** [comment]
@tdudgeon says this can be considered done. But there are some outstanding quirks and datasets.

- Some edges are dropped because they don't have corresponding nodes (missing from fragmentation data for unknown reason)
- Most complex 60M from Matteo's 250M have not been fragmented
- Old Enamine data is not included

**Comment by mwinokan** [comment]
@tdudgeon says there's about 12000 missing edges and the parent is always missing so it's hard to analyse

**Comment by mwinokan** [comment]
@tdudgeon clarifies that the chemspace data has not been combined

The Matteo data is in the dev cluster (and @mwinokan is testing in #1745)

@alanbchristie please proceed with the final combination and roll it out to the dev and production clusters

The graph-sw2 cluster should be kept alive but the old graph on the production cluster can be shut down

**Comment by alanbchristie** [comment]
Tim has confirmed that the S3 (`AWS_BUCKET_PATH`) is: -

- `datasets/14_enamine_mferla_xchem_chemspace/`

> The ENV variable is `AWS_BUCKET_PATH` but it is not on AWS, that's just the env variable name.

I will try to create two new graphs in each cluster (alongside the existing) as the initialisation will take about 2 days.

**Comment by alanbchristie** [comment]
A graph should be starting in the DEV cluster but we do not have enough **Volume** quota to create one in the Prod cluster. We currently have 4 graphs running and they each need (approximately) 4TB of disc. If we can delete the one created for investigation of the new fragments then we should be able to start one in Prod - or we just wait until Monday when we can remove the exiting Dev graph and the investigation graph.

**Comment by alanbchristie** [comment]
I've been able to free-up some space so the 5th graph (in PROD) should also be underway now...

**Comment by Waztom** [comment]
Graph DB is available in Production and is en-route to development cluster. FS issue with initialisation to dev cluster. @alanbchristie is on it and does not see any major issues.  

Need to decide what to do with the old graph DB. Will wait for @mwinokan to get online and test the new DBs.

**Comment by mwinokan** [comment]
@mwinokan says that fragment knitwork will need a refactor for proper testing #1745 

@alanbchristie says there's five graphs. They are using disk and CPU resources and there is no large urgency to remove these but it is unnecessary.

- kitchen sink (old data) x2 (dev + prod)
- MFerla (new data) x2 (dev + prod)
- mini test graph on the dev cluster for @mwinokan (@alanbchristie this one can be removed for sure)
- Graph-sw2 (kitchen sink + extra data from Steph). This is the only one being used in anger by @mwinokan

**Comment by mwinokan** [comment]
The new graphs can't be used for fragment knitwork due to missing properties. Leave graphs as they are for now
